Adopt more smart pointers in WebPageProxy (Part 7)

RupinMittal opened the pull request:
#### fbdca47e7a7aee5274bd100cf2add90f269328c3
<pre>
Adopt more smart pointers in WebPageProxy (Part 7)
<a href="https://bugs.webkit.org/show_bug.cgi?id=287202">https://bugs.webkit.org/show_bug.cgi?id=287202</a>
<a href="https://rdar.apple.com/144348285">rdar://144348285</a>

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.
7th time&apos;s the charm. (Hopefully).

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::protectedWebFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::backForwardGoToItemShared):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::requestUserMediaPermissionForFrame):
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
(WebKit::WebPageProxy::beginMonitoringCaptureDevices):
(WebKit::WebPageProxy::validateCaptureStateUpdate):
(WebKit::WebPageProxy::setShouldListenToVoiceActivity):
(WebKit::WebPageProxy::syncIfMockDevicesEnabledChanged):
(WebKit::WebPageProxy::clearUserMediaState):
(WebKit::WebPageProxy::requestMediaKeySystemPermissionForFrame):
(WebKit::WebPageProxy::protectedMediaKeySystemPermissionRequestManager):
(WebKit::WebPageProxy::clearNotificationPermissionState):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
(WebKit::WebPageProxy::cancelNotification):
(WebKit::WebPageProxy::clearNotifications):
(WebKit::WebPageProxy::didDestroyNotification):
(WebKit::WebPageProxy::endPrinting):
(WebKit::WebPageProxy::updateBackingStoreDiscardableState):
(WebKit::WebPageProxy::setOverlayScrollbarStyle):
(WebKit::WebPageProxy::setScrollPerformanceDataCollectionEnabled):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
(WebKit::WebPageProxy::hasActiveVideoForControlsManager const):
(WebKit::WebPageProxy::isPlayingVideoInEnhancedFullscreen const):
(WebKit::WebPageProxy::didPerformImmediateActionHitTest):
(WebKit::WebPageProxy::requestStorageAccessConfirm):
(WebKit::WebPageProxy::requestAttachmentIcon):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromData):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromFilePath):
(WebKit::WebPageProxy::registerAttachmentIdentifier):
(WebKit::WebPageProxy::registerAttachmentsFromSerializedData):
(WebKit::WebPageProxy::cloneAttachmentData):
(WebKit::WebPageProxy::serializedAttachmentDataForIdentifiers):
(WebKit::WebPageProxy::didInsertAttachmentWithIdentifier):
(WebKit::WebPageProxy::didRemoveAttachmentWithIdentifier):
(WebKit::WebPageProxy::resetPaymentCoordinator):
(WebKit::WebPageProxy::resetSpeechSynthesizer):
(WebKit::WebPageProxy::speechSynthesisVoiceList):
(WebKit::WebPageProxy::speechSynthesisSpeak):
(WebKit::WebPageProxy::speechSynthesisCancel):
(WebKit::WebPageProxy::speechSynthesisResetState):
(WebKit::WebPageProxy::speechSynthesisPause):
(WebKit::WebPageProxy::speechSynthesisResume):
(WebKit::WebPageProxy::setOrientationForMediaCapture):
(WebKit::WebPageProxy::gpuProcessExited):
(WebKit::WebPageProxy::protectedSpeechRecognitionPermissionManager):
(WebKit::WebPageProxy::requestSpeechRecognitionPermission):
(WebKit::WebPageProxy::requestUserMediaPermissionForSpeechRecognition):
(WebKit::WebPageProxy::createRealtimeMediaSourceForSpeechRecognition):
(WebKit::WebPageProxy::processForSite):
(WebKit::WebPageProxy::webPageIDInProcess const):
(WebKit::WebPageProxy::sendToWebPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
(WebKit::SpeechSynthesisData::protectedSynthesizer):

Canonical link: <a href="https://commits.webkit.org/290246@main">https://commits.webkit.org/290246@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/266846ba21d31354c34f783f594edc8dc9b3c33a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8880 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94341 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40116 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17141 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68850 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26513 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7113 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81096 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49201 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6858 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39223 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77215 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36467 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96170 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77712 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16791 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76888 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77013 "Found 1 new API test failure: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21446 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20039 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9686 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14019 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16549 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21860 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16290 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19741 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18071 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12121,7 +12121,7 @@ void WebPageProxy::validateCaptureStateUpdate(WebCore::UserMediaRequestIdentifie
         Ref userMediaOrigin = API::SecurityOrigin::create(request->protectedUserMediaDocumentSecurityOrigin());
         Ref topLevelOrigin = API::SecurityOrigin::create(request->protectedTopLevelDocumentSecurityOrigin());
-        uiClient().decidePolicyForUserMediaPermissionRequest(*this, *webFrame, WTFMove(userMediaOrigin), WTFMove(topLevelOrigin), request.get());
+        uiClient().decidePolicyForUserMediaPermissionRequest(*this, *webFrame, userMediaOrigin, topLevelOrigin, request.get());
```
The failure here is: https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/289889@main%20(1780)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebPageProxy.cpp-validateCaptureStateUpdate-37-1dce7b.html#EndPath (same warning for `topLevelOrigin` as well.

Do we prefer just copying it as I did which will make the analyzer happy or should I suppress the warning with a FIXME comment and a radar instead?

cdumez replied:
cc @rniwa as this looks like a false positive. Copying is sad here.

Ask @rniwa but I have a slight preference for suppressing here until we can improve static analysis.

RupinMittal replied:
I'll change it to suppress when I re-upload.

RupinMittal replied:
Fixed! And sent rdar://144557500 to @rniwa

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13997,9 +14005,9 @@ void WebPageProxy::writePromisedAttachmentToPasteboard(IPC::Connection& connecti
+    MESSAGE_CHECK_BASE(protectedPreferences()->attachmentElementEnabled(), connection);
 
-    auto updateAttachmentIcon = [&] {
+    auto updateAttachmentIcon = [&, protectedThis = Ref { *this }] {
```
I think we're missing a NOESCAPE somewhere or you're hitting a false positive. If this lambda was run asynchronously, then the `&` capture wouldn't be safe anyway.

RupinMittal replied:
Seems like this lambda is being run synchronously. I'm not sure where we'd add the NOESCAPE for the capture here.

RupinMittal replied:
@rniwa Any thoughts on what to do here? Analyzer failure: https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/290059@main%20(1828)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebPageProxy.cpp-unknown-0-e629f3.html#EndPath

cdumez replied:
I would add it on the function we're passing `updateAttachmentIcon` too.

RupinMittal replied:
Yep, good idea. I'll do that. I thought for a second that alone might fix it--but `updateAttachmentIcon` gets called on it's own (not passed anywhere) for non-mac platforms, so the analyzer will likely still flag it here. Nevertheless, I'll add it to the function we're passing it to anyways.

cdumez replied:
I would hope the analyzer wouldn't complain if `updateAttachmentIcon` gets called synchronously on its own.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14010,12 +14018,12 @@ void WebPageProxy::requestAttachmentIcon(IPC::Connection& connection, const Stri
 #if PLATFORM(MAC)
     if (RefPtr attachment = attachmentForIdentifier(identifier); attachment && attachment->shouldUseFileWrapperIconForDirectory()) {
-        attachment->doWithFileWrapper([&, updateAttachmentIcon = WTFMove(updateAttachmentIcon)] (NSFileWrapper *fileWrapper) {
+        attachment->doWithFileWrapper([&, protectedThis = Ref { *this }, updateAttachmentIcon = WTFMove(updateAttachmentIcon)] (NSFileWrapper *fileWrapper) {
```
Same comment about NOESCAPE.

RupinMittal replied:
Adding a NOESCAPE should fix this one.

cdumez commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -206,14 +206,18 @@ struct PrivateClickMeasurementAndMetadata {
 
+#if ENABLE(SPEECH_SYNTHESIS)
 struct SpeechSynthesisData {
     Ref<WebCore::PlatformSpeechSynthesizer> synthesizer;
```
Does this ever get reassigned, maybe it can be marked `const` to avoid the protected function.

RupinMittal replied:
It doesn't ever get re-assigned. It's assigned just once here: https://github.com/WebKit/WebKit/blob/fc41c7f03db40c81cb26976470ef16abbe61fff2/Source/WebKit/UIProcess/WebPageProxy.cpp#L14483.

I did try to make mark the `synthesizer` member const yesterday but got this error on the above line:

`Source/WebKit/UIProcess/WebPageProxy.cpp:14492:37: error: no viable overloaded '='`
 `14492 |     optionalSpeechSynthesisData = { PlatformSpeechSynthesizer::create(*this), nullptr, nullptr, nullptr, nullptr, nullptr };`
 
 Is there a way to lazy initialize or something? I haven't been able to get the const working here.

RupinMittal replied:
optional(718, 65): Candidate function not viable: cannot convert initializer list argument to 'nullopt_t'
optional(723, 45): Candidate function not viable: cannot convert initializer list argument to 'const optional<SpeechSynthesisData>

cdumez replied:
This seems to work:
```diff
diff --git a/Source/WebKit/UIProcess/WebPageProxy.cpp b/Source/WebKit/UIProcess/WebPageProxy.cpp
index b0e818f461c6..efe489036b3f 100644
--- a/Source/WebKit/UIProcess/WebPageProxy.cpp
+++ b/Source/WebKit/UIProcess/WebPageProxy.cpp
@@ -14479,7 +14479,7 @@ void WebPageProxy::resetSpeechSynthesizer()
 SpeechSynthesisData& WebPageProxy::Internals::speechSynthesisData()
 {
     if (!optionalSpeechSynthesisData)
-        optionalSpeechSynthesisData = { PlatformSpeechSynthesizer::create(*this), nullptr, nullptr, nullptr, nullptr, nullptr };
+        optionalSpeechSynthesisData.emplace(PlatformSpeechSynthesizer::create(*this), nullptr, nullptr, nullptr, nullptr, nullptr);
     return *optionalSpeechSynthesisData;
 }
 
diff --git a/Source/WebKit/UIProcess/WebPageProxyInternals.h b/Source/WebKit/UIProcess/WebPageProxyInternals.h
index 92042bc4c48c..85e0d8f37733 100644
--- a/Source/WebKit/UIProcess/WebPageProxyInternals.h
+++ b/Source/WebKit/UIProcess/WebPageProxyInternals.h
@@ -207,7 +207,7 @@ struct PrivateClickMeasurementAndMetadata {
 };
 
 struct SpeechSynthesisData {
-    Ref<WebCore::PlatformSpeechSynthesizer> synthesizer;
+    const Ref<WebCore::PlatformSpeechSynthesizer> synthesizer;
     RefPtr<WebCore::PlatformSpeechSynthesisUtterance> utterance;
     CompletionHandler<void()> speakingStartedCompletionHandler;
     CompletionHandler<void()> speakingFinishedCompletionHandler;
```

RupinMittal replied:
Ohhh nice. I think this should work, thank you.

RupinMittal committed:
Adopt more smart pointers in WebPageProxy (Part 7)
https://bugs.webkit.org/show_bug.cgi?id=287202
rdar://144348285

Reviewed by Ryosuke Niwa.

Smart pointer adoption as per the static analyzer.
7th time's the charm. (Hopefully).

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::protectedWebFrame):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::launchProcess):
(WebKit::WebPageProxy::backForwardGoToItemShared):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::requestUserMediaPermissionForFrame):
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
(WebKit::WebPageProxy::beginMonitoringCaptureDevices):
(WebKit::WebPageProxy::validateCaptureStateUpdate):
(WebKit::WebPageProxy::setShouldListenToVoiceActivity):
(WebKit::WebPageProxy::syncIfMockDevicesEnabledChanged):
(WebKit::WebPageProxy::clearUserMediaState):
(WebKit::WebPageProxy::requestMediaKeySystemPermissionForFrame):
(WebKit::WebPageProxy::protectedMediaKeySystemPermissionRequestManager):
(WebKit::WebPageProxy::clearNotificationPermissionState):
(WebKit::WebPageProxy::pageWillLikelyUseNotifications):
(WebKit::WebPageProxy::showNotification):
(WebKit::WebPageProxy::cancelNotification):
(WebKit::WebPageProxy::clearNotifications):
(WebKit::WebPageProxy::didDestroyNotification):
(WebKit::WebPageProxy::endPrinting):
(WebKit::WebPageProxy::updateBackingStoreDiscardableState):
(WebKit::WebPageProxy::setOverlayScrollbarStyle):
(WebKit::WebPageProxy::setScrollPerformanceDataCollectionEnabled):
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
(WebKit::WebPageProxy::hasActiveVideoForControlsManager const):
(WebKit::WebPageProxy::isPlayingVideoInEnhancedFullscreen const):
(WebKit::WebPageProxy::didPerformImmediateActionHitTest):
(WebKit::WebPageProxy::requestStorageAccessConfirm):
(WebKit::WebPageProxy::requestAttachmentIcon):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromData):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromFilePath):
(WebKit::WebPageProxy::registerAttachmentIdentifier):
(WebKit::WebPageProxy::registerAttachmentsFromSerializedData):
(WebKit::WebPageProxy::cloneAttachmentData):
(WebKit::WebPageProxy::serializedAttachmentDataForIdentifiers):
(WebKit::WebPageProxy::didInsertAttachmentWithIdentifier):
(WebKit::WebPageProxy::didRemoveAttachmentWithIdentifier):
(WebKit::WebPageProxy::resetPaymentCoordinator):
(WebKit::WebPageProxy::resetSpeechSynthesizer):
(WebKit::WebPageProxy::speechSynthesisVoiceList):
(WebKit::WebPageProxy::speechSynthesisSpeak):
(WebKit::WebPageProxy::speechSynthesisCancel):
(WebKit::WebPageProxy::speechSynthesisResetState):
(WebKit::WebPageProxy::speechSynthesisPause):
(WebKit::WebPageProxy::speechSynthesisResume):
(WebKit::WebPageProxy::setOrientationForMediaCapture):
(WebKit::WebPageProxy::gpuProcessExited):
(WebKit::WebPageProxy::protectedSpeechRecognitionPermissionManager):
(WebKit::WebPageProxy::requestSpeechRecognitionPermission):
(WebKit::WebPageProxy::requestUserMediaPermissionForSpeechRecognition):
(WebKit::WebPageProxy::createRealtimeMediaSourceForSpeechRecognition):
(WebKit::WebPageProxy::processForSite):
(WebKit::WebPageProxy::webPageIDInProcess const):
(WebKit::WebPageProxy::sendToWebPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
(WebKit::SpeechSynthesisData::protectedSynthesizer):

Canonical link: https://commits.webkit.org/290246@main
